source/tutorials/callpackage.md

```diff
@@ -0,0 +1,283 @@
+A function that takes an attribute set, with attributes corresponding to derivations in the top-level package set, and returns a derivation.
+
+```{code-block} nix
+:caption: hello.nix
+{ writeShellScriptBin }:
+writeShellScriptBin "hello" ''
+  echo "Hello, world!"
+''
+```
+
+:::{dropdown} Detailed explanation
+`hello.nix` declares a function which takes as argument an attribute set with one element `writeShellScriptBin`.
+[`writeShellScriptBin`](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeShellScriptBin) is a function that happens to exist in Nixpkgs, a [build helper](https://nixos.org/manual/nixpkgs/unstable/#part-builders) that returns a derivation.
+The derivation output in this case is an executable shell script with the contents `echo "hello world"` named `"hello"`.
+:::
+
+Now create a file `default.nix` with the following contents:
+
+```{code-block} nix
+:caption: default.nix
+let
+  pkgs = import <nixpkgs> { };
+in
+pkgs.callPackage ./hello.nix { }
+```
+
+Realise the derivation in `default.nix` and run the executable that is produced:
+
+```shell-session
+$ nix-build
+$ ./result/bin/hello
+Hello, world!
+```
+
+The argument `writeShellScriptBin` gets filled in automatically when the function in `hello.nix` is evaluated.
+For every attribute in the function's argument, `callPackage` passes an attribute from the `pkgs` attribute set if it exists.
+
+It may appear cumbersome to create the extra file `hello.nix` for the package in such a simple setup.
+We have done so because this is exactly how Nixpkgs is organised:
+Every package recipe is a file that declares a function.
+This function takes as arguments the package's dependencies.
+
+## Parameterised builds
+
+Change the `default.nix` to produce an attribute set of derivations, with the attribute `hello` containing the original derivation:
+
+```{code-block} nix
+:caption: default.nix
+let
+  pkgs = import <nixpkgs> { };
+in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+}
+```
+
+When building the attribute `hello`, by accessing it with the [`-A` / `--attr` option](https://nix.dev/manual/nix/2.19/command-ref/nix-build#opt-attr), the result will be the same as before:
+
+```shell-session
+$ nix-build -A hello
+$ ./result/bin/hello
+Hello, world!
+```
+
+Also change `hello.nix` to add an additional parameter `audience` with default value `"world"`:
+
+```{code-block} nix
+:caption: hello.nix
+{
+  writeShellScriptBin,
+  audience ? "world",
+}:
+writeShellScriptBin "hello" ''
+  echo "Hello, ${audience}!"
+''
+```
+
+This also does not change the result.
+
+Things get more interesting when changing `default.nix` to make use of this new argument.
+Pass the parameter `audience` in the second argument to `callPackage`:
+
+```{code-block} diff
+:caption: default.nix
+ let
+   pkgs = import <nixpkgs> { };
+ in
+ {
+   hello = pkgs.callPackage ./hello.nix { };
++  hello-people = pkgs.callPackage ./hello.nix { audience = "people"; };
```

Comment:
This is generally not recommended because any changes to the arguments to `hello` won't influence `hello-people`. It's also conceptually the easiest if each file is only imported from one location. This also leads to the question later when `.override` should be preferred over using another `callPackage`.

I'd say this is best avoided by just passing audience to `hello` instead:

```suggestion
-  hello = pkgs.callPackage ./hello.nix { };
+  hello = pkgs.callPackage ./hello.nix { audience = "people"; };
```

This might warrant moving the introduction of `-A` further down.